### DISCUSSION TITLE
Bintray shuts down

### DISCUSSION THREAD

**Discussion by AlekSi** [Homebrew/discussions#691] [Everyday usage] [answered]
https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/

Will it affect Homebrew's bottles?

**Comment by SMillerDev** [comment]
Yes, definitely. We're currently in talks with JFrog to figure this out. But if you have a couple of terabytes of storage laying around somewhere with unlimited bandwidth ideas are welcome 😄

**Reply by carlocab** [reply]
Bandwidth would help. Homebrew can probably afford the necessary storage. Bandwidth is expensive.

**Reply by tholu** [reply]
@SMillerDev What's the current storage and bandwidth requirements of homebrew?

**Reply by iMichka** [reply]
Around 2,5 TB of bottle storage and 450 TB/month of bandwidth (mac + linux packages).
These figures are increasing from month to month.

**Reply by slonopotamus** [reply]
You could possibly try contacting cloudsmith.io. Your storage/bandwidth requirements are way greater than what they offer by default for open source projects, but something might happen via a [custom sponsorship agreement](https://help.cloudsmith.io/docs/open-source-hosting-policy#sponsorship).

Disclaimer: I am not affiliated with cloudsmith.io in any way.

**Reply by iMichka** [reply]
Thanks for the proposal, we are already exploring some options.

We are handling this internally for the moment. To avoid having everyone contacting multiple providers and creating a lot of back and forth and miscommunication, our project leadership committee and the project leader are centralising the effort. We will let the community know once we have sorted out the issue.

**Reply by mattfarina** [reply]
Might I suggest that any solution you setup has custom homebrew DNS for the endpoint. That way if you need to move again you have more control over the situation.

**Reply by jonchang** [reply]
Yes, we are considering as well.

**Reply by pumano** [reply]
Today I found https://bunny.net is a CDN for a cheap price. But 450TB bandwidth is huge. Maybe they can help if someone from homebrew crew wrote few lines to their email address.

**Reply by crunchtime-ali** [reply]
How about approaching Apple to foot the bill? After all they are the ones who are massively profiting off of this project. 
When using GCP's Cloud Storage this would be around $620 per month.
With S3 + CloudFront it should be a tad cheaper but its way more difficult to calculate as there are volume tiers and transfer costs differ per region.

**Reply by cliffrowley** [reply]
I'm actually getting an error trying to upgrade homebrew packages now, I'm assuming it's because of this?

**Reply by gromgit** [reply]
No, Bintray's DNS wasn't resolving properly for a time. It's fixed now: https://github.com/Homebrew/brew/issues/10739

**Reply by KillDozerX2** [reply]
Are there any updates on this?

**Reply by lawgimenez** [reply]
Someone should email Apple!

**Reply by SMillerDev** [reply]
For what exactly? Apple isn't a storage/cdn provider so it would be of no use to us.

If you want updates on work homebrew maintainers are doing for this and other things, read the pull requests in homebrew/brew and homebrew/homebrew-core. We're a package manager, not a news agency.

**Reply by mikesep** [reply]
Per https://github.com/Homebrew/brew/pull/10677, it looks like Homebrew is heading toward storing bottles at the Internet Archive (archive.org).

**Reply by cromatikap** [reply]
> Bandwidth would help. Homebrew can probably afford the necessary storage. Bandwidth is expensive.

It looks like what [the permaweb](https://www.arweave.org/) is designed for. You can assume unlimited storage and bandwidth from the blockchain built-in economic incentives.

**Reply by vejmarie** [reply]
So the internet archive approach is going to be the one taken ? I am asking because I liked to get my freecad bottle keep running after bintray goes down.

**Reply by gromgit** [reply]
You can stash your bottles anywhere, as long as the `root_url` value is correctly set. That's how the [Homebrew GitHub Releases](https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/) mechanism works.

**Reply by SMillerDev** [reply]
You also don't need to use the same host as homebrew-core does for taps.  If you like internet archive and homebrew supports it, go nuts.

**Reply by MikeMcQuaid** [reply]
> Might I suggest that any solution you setup has custom homebrew DNS for the endpoint. That way if you need to move again you have more control over the situation.

It's very easy for us to push DNS changes to users so: no need to add complexity/another point of failure here.

> Someone should email Apple!

We are in contact with Apple.

> Per [Homebrew/brew#10677](https://github.com/Homebrew/brew/pull/10677), it looks like Homebrew is heading toward storing bottles at the Internet Archive (archive.org).

We investigated this but we are not taking this direction. I'll post a separate answer on what we're doing.

**Reply by Uzume** [reply]
I am assuming these are going to be a problems for the Linuxbrew people:
- https://github.com/Homebrew/linuxbrew-core/search?q=bintray.com

There also seem to be many references to Bintray in the APIs:
- https://github.com/Homebrew/formulae.brew.sh/search?q=bintray.com

An example of which can be seen here:
- https://formulae.brew.sh/api/formula/berkeley-db.json

Are the plans to address these types of issues?

**Reply by MikeMcQuaid** [reply]
https://github.com/Homebrew/homebrew-core/issues/74335

**Comment by MikeMcQuaid** [comment, marked as answer]
I can confirm that we've migrated Homebrew's binary packages to GitHub Packages: https://github.com/orgs/Homebrew/packages

This will be the default download location as-of the next Homebrew stable release: https://github.com/Homebrew/brew/pull/11070

**Reply by AlekSi** [reply]
Mike, is there any writeup on how you upload non-Docker-image files to GitHub _Container_ Registry? I'm looking to do something similar (without making a GitHub release for every commit) but can't figure out how.

**Reply by MikeMcQuaid** [reply]
@AlekSi Not yet. There may be eventually. If you check out `brew pr-upload` and `github_packages.rb` you can see what we're doing there. All the tooling we use is open source and we'll fix any bugs you encounter.

**Reply by AlekSi** [reply]
Thank you

**Reply by darkRaspberry** [reply]
Linuxbrew packages are moved?

**Reply by SMillerDev** [reply]
Yes

**Reply by darkRaspberry** [reply]
Thanks i love it.

**Reply by shrinidhi111** [reply]
Hi team!

Before I would download the bottles/binaries directly from bintray for some testing... Is that not possible now?

I am only able to see docker pulls but how do I download the binaries themselves?

**Reply by MikeMcQuaid** [reply]
@shrinidhi111 No longer possible, they have all been deleted by Bintray.

**Comment by rawlins** [comment]
Following up to some discussion [here](https://github.com/Homebrew/brew/pull/11070#issuecomment-817862600) (sorry to switch locations but I realized an open discussion topic was probably better than a closed PR):

* there's currently a brownout at bintray "to warn users about May 1st as sunset date for Bintray".
* Anyone using a 10.15 runner (i.e. `macos-latest`) with the default settings will be getting a version of homebrew that is apparently not updated to use the new download location (Homebrew 3.0.10 looks like). In the actions output the error is not very informative about what's going on, users will just see `curl: (22) The requested URL returned error: 502 Bad Gateway`. The URL itself just gives `Forbidden!`.
* It's easy to fix by updating homebrew, just was not at all apparent to me (as a casual actions user) that this was the solution without a bunch of digging into the above. (What we did in particular was explicitly use the [setup homebrew](https://github.com/Homebrew/actions/tree/master/setup-homebrew) action rather than relying on the implicit setup.)
* Unless github updates their default macos image by May 1 this will probably come back again. Edit: in https://github.com/actions/virtual-environments/issues/3165 an actions staff member indicated that `macos-latest` would probably update to the new version of homebrew mid-next-week.

Hope this is helpful in saving other people some time!

**Reply by lukeredpath** [reply]
We're struggling with this problem with our Circle CI pipelines at the moment and I'm not sure what's going wrong.

We were alerted to this issue by the Bintry brown-out - we were not routinely auto-updating Homebrew (we disable auto-update when we install) and I don't know how often Circle CI update their images. As a temporary fix I tried re-enabling auto-update and I can see in our build logs that it is running a `brew update` but it still fails to install the package `xclogparser`, with the following output:

```
brew install xclogparser
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 3 taps (homebrew/cask-versions, homebrew/core and adoptopenjdk/openjdk).
==> New Formulae
as-tree                    kubergrunt                 python-tabulate
bas55                      latino                     python-tk@3.9
brook                      leaf-proxy                 qt-libiodbc
cadence-workflow           libmd                      qt-mariadb
cherrytree                 libpipeline                qt-mysql
clusterctl                 libprelude                 qt-percona-server
coin3d                     libunwind                  qt-postgresql
crackpkcs                  linux-pam                  qt-unixodbc
crispy-doom                lttng-ust                  rdkit
curlie                     mapcidr                    rosa-cli
dasel                      moto                       rtl_433
datalad                    multitime                  saml2aws
delve                      mx                         smu
djl-serving                name-that-hash             snowpack
dua-cli                    net-tools                  spaceship
enkits                     numactl                    sqlancer
enzyme                     obfs4proxy                 systemd
epr                        open-adventure             threemux
geph4                      openmama                   timg
ghc@9                      openmodelica               tomcat@9
glibc                      oras                       truffle
go-boring                  orgalorg                   tz
go@1.15                    pandoc-plot                virtualenv
gopass-jsonapi             pcalc                      waffle
haruhi-dl                  pcp                        wllvm
haskell-language-server    phpbrew                    xcprojectlint
iconsur                    projectm                   xray
influxdb@1                 pure                       yubikey-agent
kertish-dfs                pyqt-3d                    zlib-ng
kickstart                  pyqt-builder               zsh-vi-mode
klee                       pyqt-networkauth
kotlin-language-server     pyside@2
==> Updated Formulae
Updated 5105 formulae.
==> Renamed Formulae
glibmm@2.64 -> glibmm@2.66
ht-rust -> xh
kde-extra-cmake-modules -> extra-cmake-modules
kde-karchive -> karchive
kde-kdoctools -> kdoctools
kde-ki18n -> ki18n
kde-threadweaver -> threadweaver
libsasl2 -> cyrus-sasl
minizip2 -> minizip-ng
pangomm@2.42 -> pangomm@2.46
pyqt5 -> pyqt@5
qt5 -> qt@5
==> Deleted Formulae
atlassian-cli              avian                      geant4
==> New Casks
ableton-live-intro10       adoptopenjdk16-openj9      dnagedcom-beta
ableton-live-lite10        adoptopenjdk16-openj9-jre  local-beta
ableton-live-standard10    arduino-beta               netnewswire-beta
ableton-live-suite10       clashx-pro                 olympus
adoptopenjdk16             dash5                      qq-beta
adoptopenjdk16-jre         deadbeef-nightly           zulu15
==> Updated Casks
Updated 74 casks.
==> Deleted Casks
hex-fiend-beta

==> Downloading https://homebrew.bintray.com/xclogparser-0.2.26.catalina.bottle.
##O#- #                                                                       
curl: (22) The requested URL returned error: 403 Forbidden
Trying a mirror...
==> Downloading https://ghcr.io/v2/homebrew/core/xclogparser-0.2.26.catalina.bot
==> Downloading from https://github.com/-/v2/packages/container/package/homebrew
#=#=#                                                                         
curl: (22) The requested URL returned error: 404 
Error: Failed to download resource "xclogparser"
Download failed: https://ghcr.io/v2/homebrew/core/xclogparser-0.2.26.catalina.bottle.tar.gz

Exited with code exit status 1
CircleCI received exit code 1
```

If I SSH into the Circle CI worker after it fails I can see it has updated to brew 3.1.1. I can install this package on my local machine using brew 3.1.1. 

Does anybody know how to fix this?

**Reply by lukeredpath** [reply]
So it looks like an explicit `brew update` did the trick, I'm not sure why that is any different. Sadly it adds nearly 5 minutes overhead to our builds so its not a sustainable solution, we may need to raise a support ticket with Circle CI for this.

**Comment by mikezs** [comment]
Just in case anyone's having more problems getting an update to properly be forced, I'm using this script:

    cd $(brew --repo); git fetch; git reset --hard origin/master
    cd "$(brew --repo)/Library/Taps/homebrew/homebrew-core"; git pull https://github.com/Homebrew/homebrew-core
    cd "$(brew --repo)/Library/Taps/homebrew/homebrew-cask"; git pull https://github.com/Homebrew/homebrew-cask

**Reply by carlocab** [reply]
You can also try `brew update-reset`. (Make sure to stash any changes you've made locally that you want to keep.)

**Comment by maltfield** [comment]
In-case anyone else ended-up in this thread trying to figure out how to download files from GitHub Packages. This no longer works:

```
wget https://homebrew.bintray.com/bottles/wget-1.20.3_2.catalina.bottle.tar.gz
```

Unfortunately, doing this equivalent in GitHub Packages is *extremely* difficult. I spent a few weeks figuring it out, and I wrote a long article that describes how to download brew bottles from GitHub Packages [here](https://tech.michaelaltfield.net/2024/09/03/container-download-curl-wget/)

 * https://tech.michaelaltfield.net/2024/09/03/container-download-curl-wget/

**Reply by p-linnane** [reply]
You can use `brew fetch` to grab bottles in an easier fashion. You can also add `--bottle-tag` to grab for a different OS/architecture. They'll be in `$(brew --cache)`.

**Reply by Bo98** [reply]
Note that for Homebrew packages you don't need a valid auth token. You can pass `Authorization: Bearer QQ==`

**Reply by maltfield** [reply]
@p-linnane that doesn't work on linux (eg TAILS). This is needed to verify the files on a non-mac system, since Brew doesn't have any built-in cryptographic authenticity checks for packages that it downloads. For more info, [see the `Why?` section of the article](https://tech.michaelaltfield.net/2024/09/03/container-download-curl-wget/#why) describing 3TOFU.

 * https://tech.michaelaltfield.net/2024/09/03/container-download-curl-wget/#why

**Reply by p-linnane** [reply]
I'm not sure what you mean by it not working for Linux. Here's me pulling down a bottle for `hello` for macOS Sonoma while on Linux.
```
linuxbrew@codespaces-a2cb05:~/.linuxbrew/Homebrew$ uname -a
Linux codespaces-a2cb05 6.5.0-1022-azure #23~22.04.1-Ubuntu SMP Thu May  9 17:59:24 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

linuxbrew@codespaces-a2cb05:~/.linuxbrew/Homebrew$ brew fetch --bottle-tag arm64_sonoma hello
==> Downloading https://ghcr.io/v2/homebrew/core/hello/manifests/2.12.1
Already downloaded: /home/linuxbrew/.cache/Homebrew/downloads/1ea4c3c37828c1113de2100e5188b954c72e9c25fcbc67c76cc55cfeec2d4ab6--hello-2.12.1.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/hello/blobs/sha256:c1770a992b197e450da7cf2282fdeaae855765f66609063bfb963ffb2ccf69a5
############################################################################################################################################################################################################### 100.0%
Downloaded to: /home/linuxbrew/.cache/Homebrew/downloads/6211799d7628b0a9502167928b43392eee5bbef99177ee72cc02cb04ab986ee5--hello--2.12.1.arm64_sonoma.bottle.tar.gz
SHA256: c1770a992b197e450da7cf2282fdeaae855765f66609063bfb963ffb2ccf69a5

linuxbrew@codespaces-a2cb05:~$ ls $(brew --cache)/downloads
1ea4c3c37828c1113de2100e5188b954c72e9c25fcbc67c76cc55cfeec2d4ab6--hello-2.12.1.bottle_manifest.json
6211799d7628b0a9502167928b43392eee5bbef99177ee72cc02cb04ab986ee5--hello--2.12.1.arm64_sonoma.bottle.tar.gz
```

`brew` checks bottle checksums to ensure they haven't been tampered with. Here's an example where I modified the bottle hash in a formula:

```
╰─ brew fetch --bottle-tag arm64_sonoma hello
==> Downloading https://ghcr.io/v2/homebrew/core/hello/manifests/2.12.1
Already downloaded: /Users/patrick/Library/Caches/Homebrew/downloads/1ea4c3c37828c1113de2100e5188b954c72e9c25fcbc67c76cc55cfeec2d4ab6--hello-2.12.1.bottle_manifest.json
Error: Failed to download resource "hello_bottle_manifest"
Couldn't find manifest matching bottle checksum.
```

**Reply by maltfield** [reply]
@p-linnane is it possible to securely install brew on TAILS?

> brew checks bottle checksums to ensure they haven't been tampered with

This is concerning. **Verifying checkums absolutely does not check for malicious tampering** (authenticity). Verifying checksums merely verifies that the file wasn't corrupt (integrity). If someone is doing something malicious, then verifying a checksum would not catch it (unless the malicious actor was exceedingly dumb and didn't modify the checksum too). 

In order to verify that the checksum hasn't been tampered-with, you'd need to verify a signature. Brew does not verify signatures. Everything that brew downloads is at-risk of being malicious. For more info, see [3TOFU](https://tech.michaelaltfield.net/2024/08/04/3tofu/)

**Reply by p-linnane** [reply]
I'm not sure what your definition of 'securely install' is. The only methods for linux are using the bash script or un-taring into a directory, though the latter is not supported.

I'm not quite sure you understand how Homebrew works. We take source code provided by an upstream and build it. Then we distribute the binaries as bottles. What exactly are you concerned will be tampered with?

We also now generate attestations when we publish bottles, which you can find [here](https://github.com/Homebrew/homebrew-core/attestations) and enable by setting `HOMEBREW_VERIFY_ATTESTATIONS=1`.

**Reply by MikeMcQuaid** [reply]
> Everything that brew downloads is at-risk of being malicious.

@maltfield This is a strong statement to be throwing around. If you have reproducible vulnerabilities and you're a security professional you will be aware that random comments on a discussion board is not a serious way of disclosing them. Note that @p-linnane is a working security professional and member of Homebrew's security team. We also recently had a professional security external audit done by a respected member in this space: https://brew.sh/2024/07/30/homebrew-security-audit/

**Reply by maltfield** [reply]
> I'm not sure what your definition of 'securely install' is

@p-linnane it's quite simple: installing where the release is verified with a cryptographic signature. Debian apt repos do this by default. Many appimage developers sign their releases. Downloading a tarball could work -- if that tarball was crypographically signed.

See also:

 * https://apple.stackexchange.com/questions/394401/cryptographic-integrity-and-authentication-validation-of-homebrew-installer

@MikeMcQuaid I'm a bit confused, and I'm quite surprised that the brew maintainers are not aware of these existing vulnerabilities in the design of your own system. My understanding was that you were aware of the issues, but you made an intentional choice to deprioritize them. But to actually not be aware of the issues is, well, alarming.

Yes, Brew provides no protection against publishing infrastructure compromise. Brew provides no protection against an active MITM attack (https can be defeated by a number of ways, such as [having one (subordinate) CA in a root store](https://security.stackexchange.com/questions/234052/where-can-i-find-a-list-of-all-government-agencies-with-cas-in-pki-root-stores) or by DNS compromise). Historically, all of these things have happened to various orgs, including many open-source projects like Linux Mint and Monero.

 * https://github.com/cncf/tag-security/tree/main/community/catalog/compromises

In the case of Linux Mint and Monero, their releases were cryptographically signed. So anyone who checked the signatures was able to see that the release was tampered-with.

In the case of Brew, if someone compromises the publishing infrastructure (or DNS or sets up an active MITM attack against a brew user, etc), then the attacker can modify the hash just as easily as they can modify the release. This is why it doesn't protect against malicious attacks. Hash verification is simply to check for corruption.

**Reply by slonopotamus** [reply]
> Brew provides no protection against publishing infrastructure compromise.

Are you talking about compromising [homebrew-core](https://github.com/Homebrew/homebrew-core) git repo or bottle artifacts? I believe the latter is verified against SHA256 hashes in formula files ([example](https://github.com/Homebrew/homebrew-core/blob/master/Formula/b/b2-tools.rb#L10-L18)). Thus, hashes and files come from different sources and attacker would need to tamper _both_ of them.

**Reply by p-linnane** [reply]
If you are in a position where you believe a threat actor is trying to MITM you, change bottle hashes, and get you to install a malicious bottle, then you should not be using Homebrew. Most users retrieve bottle information via our signed JSON API. Users can also opt into our attestation beta by setting the environment variable I previously provided. We are responsible for the contents of Homebrew/core. We are not responsible for the security of your HTTPS connections.

This is a project run entirely by a team of less than 30 volunteers for free. Read the BSD-2-Clause license and then open a pull request if you feel you can implement the feature you're describing.

**Reply by Bo98** [reply]
To be clear here: we don't blindly accept any checksum. The checksums themselves are contained in `https://formulae.brew.sh/api/formula.jws.json` which is signed by a certificate that's pinned in `brew` itself: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/api/homebrew-1.pem.

**Reply by woodruffw** [reply]
@maltfield can you walk through, step by step, where you think Homebrew is missing or has insufficient authenticity and/or integrity guarantees? For Homebrew's part, here are the things that Homebrew does and the properties they're expected to provide:

1. Homebrew is installed and updated over HTTPS only. We assume the security of your local CA store and the integrity of the Web PKI, similar to every other major application-level user of HTTPS. This ensures confidentiality, authenticity, and integrity at the transport level.
2. To retrieve bottle metadata, Homebrew makes HTTPS connections to a JSON API. Homebrew's JSON API responses are **additionally** digitally signed with a key, meaning that the hashes we verify bottles against are digitally signed over, both at the transport layer (HTTPS) and at the application layer (`brew`), with the latter being signed by a non-PKI key that only Homebrew's maintainers control.
3. The bottle hashes returned from the JSON API are used to pull bottles from the CDN, again over HTTPS.
4. The bottles pulled from CDN are verified against their expected hashes, providing our baseline application-level integrity guarantee.
5. Optionally, beyond everything else, you can set `HOMEBREW_VERIFY_ATTESTATIONS=1` to verify a bottle's *provenance*, which *additionally* establishes the bottle's integrity and authenticity modulo the Homebrew CI/CD's cryptographic identity.

(1) through (4) are hard invariants of a normal Homebrew installation. (5) is currently optional, but will become a default once we have a more stable implementation of attestation verification.

TL;DR: even if (and this is a **huge** if) an attacker somehow MITMs you via a rogue CA or trust store implant, Homebrew's JSON API still independently signs bottle metadata. Moreover, enabling attestation verification will preclude the possibility of an attacker somehow obtaining our JSON API signing key, since bottle provenance is generated entirely with short-lived keys tied to CI/CD identities.

**Reply by MikeMcQuaid** [reply]
> I'm a bit confused, and I'm quite surprised that the brew maintainers are not aware of these existing vulnerabilities in the design of your own system. My understanding was that you were aware of the issues, but you made an intentional choice to deprioritize them. But to actually not be aware of the issues is, well, alarming.

We are/I am aware people (including you) consider checksums to be essentially worthless for security but signing to be the solution to the problem.

I have not seen evidence of an attack that can be achieved on our current infrastructure that would be avoided by signing.

> Are you talking about compromising [homebrew-core](https://github.com/Homebrew/homebrew-core) git repo or bottle artifacts? I believe the latter is verified against SHA256 hashes in formula files ([example](https://github.com/Homebrew/homebrew-core/blob/master/Formula/b/b2-tools.rb#L10-L18)). Thus, hashes and files come from different sources and attacker would need to tamper _both_ of them.

Yes, this.

> Yes, Brew provides no protection against publishing infrastructure compromise.

I disagree. You would need to both compromise GitHub Packages and generate a malicious artefact with an identical SHA-256 checksum. Checksums are stored in separate infrastructure from 

> So anyone who checked the signatures was able to see that the release was tampered-with.

Unless you also MITMd both the signature store and the resources signed by the signatures. 

> then the attacker can modify the hash just as easily as they can modify the release

They would have to compromise both the GitHub Git infrastructure and the GitHub Packages infrastructure to do this.

It's hard to see a world in which being able to compromise both of these but somehow the signing infrastructure we use is impenetrable to that attack.

**Reply by maltfield** [reply]
> If you are in a position where you believe a threat actor is trying to MITM you, change bottle hashes, and get you to install a malicious bottle, then you should not be using Homebrew. 

Yeah, that's pretty concerning. I think that we should be building tools that are capable of protecting against such attacks. It's not hard. `apt-secure` has been enforced since, what, 2005?

**Reply by maltfield** [reply]
> To be clear here: we don't blindly accept any checksum. The checksums themselves are contained in `https://formulae.brew.sh/api/formula.jws.json` which is signed by a certificate that's pinned in `brew` itself: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/api/homebrew-1.pem.

@Bo98 This is promising. Signed hashes do provide authenticity. Can you please link me to the brew documentation that describes this hash signing process?

**Reply by p-linnane** [reply]
> Yeah, that's pretty concerning. I think that we should be building tools that are capable of protecting against such attacks. It's not hard. `apt-secure` has been enforced since, what, 2005?

I agree. Go ahead and open a pull request to implement your proposed design and we are happy to review it.

**Reply by maltfield** [reply]
@woodruffw https does not provide any of these things. It's extremely weak and provides zero authenticity to literally millions of users every day. It's alarming that you would write it off as "good enough", while most other package managers just implement cryptographic signatures to provide real authenticity.

To quote directly from the link [above](https://security.stackexchange.com/questions/234052/where-can-i-find-a-list-of-all-government-agencies-with-cas-in-pki-root-stores)

> There are several "root stores" that maintain a list of trusted root CAs. These root stores are imported and used by thousands (millions?) of apps to determine which [https certificates][1] are trusted. Their contents has a *huge* impact on the entire [X.509][2] security infrastructure.
> 
> A few examples of such trusted root stores include those maintained by the following organizations:
> 
> 1. [Mozilla][4]
> 1. [Microsoft][5]
> 1. [Apple][6]
> 
> Most of the CAs in these root stores are private corporations that issue certificates for customers, such as [Verisign][7]. Depending on the policy of the root store maintainer (see links above), the list may also contain CAs controlled by nonprofit organizations or government agencies.
> 
> In the past, the maintainers of these root stores have removed CA's root certs for corporations who had lost control of their root certificates. <sup>[[8][8]][[9][9]]</sup> Similarly, there have been several historic controversies with CAs operated by government agencies. <sup>[[10][10]][[11][11]]</sup> And also with private entities that have been accused of being malicious actors (cyber-mercenaries) on behalf of governments. <sup>[[12][12]]</sup>

  [1]: https://en.wikipedia.org/wiki/Transport_Layer_Security
  [2]: https://en.wikipedia.org/wiki/X.509
  [3]: https://en.wikipedia.org/wiki/Public_key_infrastructure
  [4]: https://wiki.mozilla.org/CA
  [5]: https://aka.ms/RootCert
  [6]: https://www.apple.com/certificateauthority/ca_program.html
  [7]: https://en.wikipedia.org/wiki/Verisign
  [8]: https://en.wikipedia.org/wiki/DigiNotar#Issuance_of_fraudulent_certificates
  [9]: https://en.wikipedia.org/wiki/NortonLifeLock#Sustaining_digital_certificate_security
  [10]: https://bugzilla.mozilla.org/show_bug.cgi?id=511380
  [11]: https://bugzilla.mozilla.org/show_bug.cgi?id=1567114
  [12]: https://www.eff.org/deeplinks/2019/02/cyber-mercenary-groups-shouldnt-be-trusted-your-browser-or-anywhere-else

**Reply by maltfield** [reply]
> It's hard to see a world in which being able to compromise both of these but somehow the signing infrastructure we use is impenetrable to that attack.

@MikeMcQuaid The release signing key's private key should be stored offline; that's why it provides strong protection.

At scale you can use an M-of-N HSM system like [Fero](https://github.com/coreos/fero), but most orgs the size of Brew usually just have one or two dedicated release engineers who hold the keys locked-down on their own computer (ideally inside an HSM like a USB nitrokey or yubikey)

**Reply by p-linnane** [reply]
@maltfield I doubt you understand what it takes to run a project of this scale. Thus far you've told us all the things we're doing wrong in your eyes, while promoting your blog and failing to offer any sort of code to remedy your perceived issue. If you want your opinion to hold weight, contribute rather than criticize. This is my last communication on this topic.

**Reply by maltfield** [reply]
My contribution to the brew community was to write a very detailed article describing how to download brew packages from GitHub Packages on 3 distinct systems (for 3TOFU) to verify packages that are not signed by brew (very valuable for devs like me who need to install dependencies for our macOS builds; I don't use macOS myself).

This *is* my contribution to the community. The article is licensed CC BY-SA. I pay for the hosting myself (you're welcome), but you can also copy it to the official brew documentation if you'd like.

**Reply by woodruffw** [reply]
@maltfield Again: we have cryptographic signatures. My list includes two distinct, non-HTTPS sources of cryptographic authenticity and integrity in Homebrew. 

I don't think this is the right venue to dispel persistent misunderstandings about the security properties of HTTPS, so I'll leave only a short comment: the security of the Web PKI can be modeled and quantified like any other key distribution scheme, and is *markedly* better in 2024 in *every single way* than an ad-hoc PKI derived from PGP or similar would be. I mean that without exaggeration: in terms of key control, issuer auditability, revocation, key strength requirements, certificate transparency, root program auditability, etc.

If you don't trust the default root store installed on your system, it's incumbent upon you to change it and/or modify your trust policies to limit the (exceedingly small) likelihood that a nation state is going to run a covert intermediate CA MITM operation on you. Homebrew won't do that for you, because there's no coherent threat model for us in that situation. But again, **none of this** changes the fact that Homebrew **does in fact** perform digital signatures over its bottles.

**Reply by maltfield** [reply]
@woodruffw can you please link me to the brew documentation that describes how you utilize cryptographic signatures to ensure the authenticity of everything that `brew` downloads?

**Reply by woodruffw** [reply]
@maltfield It's not documented, since it's not a part of the public interface to `brew` -- authenticity is an application-level property of `brew`'s design. But here are some source references for you to read:

* Verification of the JSON API's JWS signature: https://github.com/Homebrew/brew/blob/9659b571778864d99cb2315348ed1248848278e8/Library/Homebrew/api.rb#L153-L177
* Attestation verification (this currently uses `gh attestation` under the hood, which is why it's not enabled by default): https://github.com/Homebrew/brew/blob/9659b571778864d99cb2315348ed1248848278e8/Library/Homebrew/attestation.rb

The first bullet is probably the closest thing to what you're thinking of in terms of digital signatures over bottles -- the JWS being verified is signing over the entire JSON API response, which includes the bottle hashes.

**Reply by Bo98** [reply]
> This is promising. Signed hashes do provide authenticity. Can you please link me to the brew documentation that describes this hash signing process?

Unfortunately I don't think the documentation exists yet. When I implemented API signing I didn't have the immediate time to do a write-up about it. It would however make sense for such documentation to exist. When attestations leave beta stage, we could do a document describing all the steps brew does from start to finish.

**Reply by carlocab** [reply]
Documentation PRs are welcome too.

**Reply by maltfield** [reply]
@Bo98 thanks. I created a ticket (for the documentation) here:

 * https://github.com/Homebrew/brew/issues/18248

That's probably also a more appropriate place to continue this discussion than this thread.

**Reply by maltfield** [reply]
Well @MikeMcQuaid just closed the ticket above. I don't understand. And he locked the thread so that I can't ask where I should re-open it.

@MikeMcQuaid I don't think this is the proper place to discuss these issues. Where *should* I open a bug report for this missing documentation?

**Reply by MikeMcQuaid** [reply]
@maltfield Please do not create any more Homebrew tickets as you have stated, despite many people engaging with you in good faith here, that you both do not use Homebrew and wish to make bold statements like "most security experts don't trust brew".

Please reread our Code of Conduct and ensure that all future communication is in keeping with it: https://github.com/Homebrew/.github/blob/master/CODE_OF_CONDUCT.md

**Reply by MikeMcQuaid** [reply]
> Where _should_ I open a bug report for this missing documentation?

It's not a bug so: do not open a bug report. You have requested it here. That is sufficient. We have and will consider your request.

**Reply by maltfield** [reply]
@MikeMcQuaid is this how you treat contributors?

I currently avoid using homebrew because of these security issues. I would *like* to use homebrew, but I need to be able to work with your team to address these issues before I can even begin to consider using `brew` (actually I do use it in my CI pipelines, but only with the 3TOFU method that I documented and published at the top of this thread).

What you're saying doesn't make sense. I've reported over 30 bugs to various software projects on GitHub related to release signing. Most of that software I don't use because I need the signing as a prereq to being able to use the software :confused: 

How the heck are we supposed to contribute bugs for prereqs that are not met??

**Reply by MikeMcQuaid** [reply]
> @MikeMcQuaid is this how you treat contributors?

Homebrew's definition of "contributors" involves people who have submitted a pull request to one of our repositories. I've been working on open source for 20 years, on Homebrew for 15 years and spent 10 years at GitHub so my views here are not uninformed.

> I would _like_ to use homebrew, but I need to be able to work with your team to address these issues

The best person to address your issues using Homebrew is you rather than us. Millions of other people are able to use Homebrew without us making the changes you want us to. Making the changes you want us to takes time away from addressing the other issues. We don't keep issues claiming to be "bugs" around open on our issues tracker when we don't have significant other numbers of people who also need the same thing.

> What you're saying doesn't make sense.

It makes sense to me and the other folks that work on Homebrew. In our space: if you think we don't make sense and you do: that's something on you to address, not us.

> Most of that software I don't use because I need the signing as a prereq to being able to use the software 😕

My prediction is that not all of that software will make the changes that you "need" because they provide their software as-is and their [maintainers owe you nothing](https://mikemcquaid.com/open-source-maintainers-owe-you-nothing/).

> How the heck are we supposed to contribute bugs for prereqs that are not met??

We don't think you are.

Again, please reread our Code of Conduct and ensure that all future communication is in keeping with it: [Homebrew/.github@master/CODE_OF_CONDUCT.md](https://github.com/Homebrew/.github/blob/master/CODE_OF_CONDUCT.md).

Your communication in this thread should be improved in future comments.

**Reply by maltfield** [reply]
Right, so, back to what I said at the top: Brew is aware of these issues. You're just de-prioritizing security.

Sounds like you have some serious leadership issues :facepalm: I wish the rest of the team the best of luck

**Reply by MikeMcQuaid** [reply]
> Right, so, back to what I said at the top: Brew is aware of these issues. You're just de-prioritizing security.

The only "issue" I see here is that we don't have documentation for something that you see as a prerequisite for using Homebrew that others don't.

Note: we have many other issues that do not warrant being included in the issue tracker. That's how we use our issue tracker. It's different to other projects.

We are not de-prioritising security. We have recently had and addressed issues raised by a professional security audit (our second). We have a security team made up with two people who professionally work in security (which you do not seem to).

> Sounds like you have some serious leadership issues 🤦 I wish the rest of the team the best of luck

This is an ad-hominem attack. You've been pointed to our Code of Conduct twice and don't seem to be acknowledging that, expressing gratitude for the time that's been put into explaining things to you or apologising when you've been showed to be incorrect or your communication has been suboptimal.

I will be more explicit in the home it warrants a response: if you continue to double-down on this behaviour and refusing to apologise or adjust: you will be blocked from the Homebrew organisation. This is your first, and depending on the response may be your last, formal warning.
